AWS EBS: Remove the attached volumes cache

justinsb opened the pull request:
There are known issues with the attached-volume state cache that just aren't
possible to fix with the current interface.

Replace it with a map of the active attach jobs (that was the original
requirement, to avoid a nasty race condition).

This costs us an extra DescribeInstance call on attach/detach, but that
seems worth it if it ends this class of bugs.

Fix #15073

justinsb committed:
AWS EBS: Remove the attached volumes cache

There are known issues with the attached-volume state cache that just aren't
possible to fix with the current interface.

Replace it with a map of the active attach jobs (that was the original
requirement, to avoid a nasty race condition).

This costs us an extra DescribeInstance call on attach/detach, but that
seems worth it if it ends this class of bugs.

Fix #15073

jsafrane commented:
I like this approach, the code looks solid too. Still, I have one observation: sometimes AWS can't attach a volume and returns error instead. So, the volume will never reach "attached" state and it will remain in `self.attaching` map forever. IMO it would be nice to check if the volume state is attaching/detaching and do some cleanup if not. I don't remember how is the state called, it happens really seldom.

Alternatively, if the volume is "attaching" for ~10 minutes, we should force-detach it and try again.

jsafrane commented:
Just FYI, I tested it today on a busy AWS node (attach/detach times out regularly) and it fixed `Error attaching EBS volume: VolumeInUse: vol-abcd1234 is already attached to an instance` for me. Kubelet  eventually found all attached volumes and started all pods.

justinsb commented:
That's great @jsafrane ...  I agree with your concern about the failure case - I think we are better off with this approach than we were before, but we are still not perfect when we time out.  I think the problem case is that an Attach operation takes longer than the timeout, as we (deliberately) leave it in the attaching map after this PR.  It will be removed if a later Attach or Detach operation in the cloud provider eventually succeeds.  If we never call Attach or Detach on it again (perhaps the pod is deleted), then we have a problem:  We have then leaked the attachment in the `attaching` map, and likely the volume is eventually attached to the node.  I think we're generally looking at a better reconciliation loop for volumes in 1.3, which would recover from this (and also failed nodes in general, cc @saad-ali ). 

My opinion is that what we've got here should eliminate a lot of issues we've seen with this cache, at the cost of a few more API calls.  It's still not perfect, but I do feel much more confident in it than the cached map when it comes to these timeout/failure scenarios.

I do wonder if we can tweak the find-attachment operation to also reconcile the attach operations, but honestly I think the 1.2 window is closed/closing.  There's always 1.2.1 !

jsafrane commented:
I agree, AWS cloud provider is in much better shape with this patch than without it. I'd really love this patch in 1.2.
LGTM

There are some corner cases we need to work on, but I think we can do it in separate PRs later.

swagiaal commented:
@justinsb I like this PR a lot! And it will definitely put us in a much better position.

> I think the problem case is that an Attach operation takes longer than the timeout, as we (deliberately) leave it in the attaching map after this PR. It will be removed if a later Attach or Detach operation in the cloud provider eventually succeeds.

That makes sense I think and it works to reserve the device in case another attach call comes in while the long attach is still working. It will get reconciled with a warning logged which sounds good to me.

Perhaps to address @jsafrane's concern we could call 'endAttaching' in error cases as well

swagiaal commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1280,8 +1280,8 @@ func (c *AWSCloud) AttachDisk(diskName string, instanceName string, readOnly boo
 	defer func() {
-		if !attached {
-			awsInstance.releaseMountDevice(disk.awsID, mountDevice)
+		if attached {
```
Maybe change this to 'if attached || errorOccured' and set 'errorOccured' to true when 'ec2.AttachVolume()' returns an error

swagiaal commented:
> Perhaps to address @jsafrane's concern we could call 'endAttaching' in error cases as well

LGTM otherwise BTW

justinsb committed:
AWS volumes: Release disk from attaching map on error

If AWS gives us an actual error (vs just timing out), we know the disk
did not attach, and so we can remove it immediately from the attaching
map.

justinsb commented:
> Perhaps to address @jsafrane's concern we could call 'endAttaching' in error cases as well

You're (both) absolutely right ... I totally missed that case.  When EC2 actually gives us an error, we know the attachment failed, so we should remove it from the attaching map.  Added a commit to cover that case.

swagiaal commented:
@justinsb thanks! LGTM
@saad-ali do you think we can get this merged ?
I will backport this to openshift either way but it would be nice if it is merged upstream.
